frontend/src/lib/components/PropertyFilters/propertyFilterLogic.ts

```diff
@@ -1,74 +1,158 @@
 import { kea } from 'kea'
 
 import { propertyFilterLogicType } from './propertyFilterLogicType'
-import { AnyPropertyFilter, EmptyPropertyFilter, PropertyFilter } from '~/types'
-import { isValidPropertyFilter, parseProperties } from 'lib/components/PropertyFilters/utils'
+import { AnyPropertyFilter, AndOrPropertyFilter, EmptyPropertyFilter, PropertyFilter } from '~/types'
+import { isValidPropertyFilter, parseProperties, parsePropertyGroups } from 'lib/components/PropertyFilters/utils'
 import { PropertyFilterLogicProps } from 'lib/components/PropertyFilters/types'
+import { AndOr } from '../MatchPropertyFilters/MatchPropertyFilters'
+
+export function isAndOrPropertyFilter(
+    filter: AndOrPropertyFilter | AnyPropertyFilter[]
+): filter is AndOrPropertyFilter {
+    return (<AndOrPropertyFilter>filter).groups !== undefined
+}
 
 export const propertyFilterLogic = kea<propertyFilterLogicType>({
     path: (key) => ['lib', 'components', 'PropertyFilters', 'propertyFilterLogic', key],
     props: {} as PropertyFilterLogicProps,
     key: (props) => props.pageKey,
 
     actions: () => ({
-        update: true,
+        update: (propertyGroupIndex?: number) => ({ propertyGroupIndex }),
         setFilter: (
             index: number,
             key: PropertyFilter['key'],
             value: PropertyFilter['value'],
             operator: PropertyFilter['operator'],
             type: PropertyFilter['type'],
-            group_type_index?: PropertyFilter['group_type_index']
-        ) => ({ index, key, value, operator, type, group_type_index }),
+            group_type_index?: PropertyFilter['group_type_index'],
+            propertyGroupIndex?: number,
+            propertyIndex?: number
+        ) => ({ index, key, value, operator, type, group_type_index, propertyGroupIndex, propertyIndex }),
         setFilters: (filters: AnyPropertyFilter[]) => ({ filters }),
-        remove: (index: number) => ({ index }),
+        remove: (propertyGroupIndex: number, propertyIndex?: number) => ({ propertyGroupIndex, propertyIndex }),
+        addFilterGroup: true,
+        addPropertyToGroup: (propertyGroupIndex) => ({ propertyGroupIndex }),
+        removeFilterGroup: (filterGroup: number) => ({ filterGroup }),
     }),
 
     reducers: ({ props }) => ({
         filters: [
-            props.propertyFilters ? parseProperties(props.propertyFilters) : ([] as AnyPropertyFilter[]),
+            props.propertyFilters
```

Comment:
@Twixes @mariusandra I need some help with updating the typing with the property logic to handle the new and/or filtering :'( What's the best way to go about this?

The TLDR of the typing issues is coming from me trying to handle two different types of filters: `properties` and `property_groups` with the same propertyLogic : (

Reply:
Because we want to handle and/or property filtering, we're now dealing with a property_groups payload of this: https://github.com/PostHog/posthog/pull/8012

Reply:
Because property_groups are at a level higher than propertyFilter, and not every place with a property filter needs property groups, how about we break the two instead of keeping it one?

A `PropertyGroupFilterLogic` manages the grouping, and the "add group" button, and has multiple keyed `PropertyFilterLogic`s, each one for a respective AND or OR group, which is managed by the `PropertyFilter`, having pure properties.

That should hopefully simplify typing quite a bit? As:

 ```
 PropertyGroupFilterType = {
     type: PropertyOperatorType,
     groups: PropertyFilter[]
 }
 ```
 
 You don't need infinite recursion on the frontend, we only support 2 levels deep, and we should make that explicit here, using the types. The above is the inner type.
 
 And for the outer, when it exists: (names tbd, just a rough example):
 
 ```
 PropertyGroupOuterType: {
    type: PropertyOperatorType,
    groups: PropertyGroupFilterType[] (this is the type above^^)
```

There will still be some wonkiness in the groupLogic, since it can be a `PropertyGroupOuterType | PropertyGroupFilterType`. But maybe we can also split the functions dealing with each, since (I think/expect) the code would be separate enough. Maybe.

Reply:
Yeah I was thinking about that too, I'll give that a try and see if it's easier than having to update all the typing everywhere that uses `<PropertyFilters` and `propertyFilterLogic`